File: slack_channel/get_current_karma_reasons_event_handler.py
import os

from commands import GetCurrentKarmaReasonsCommand
from model.karma import KarmaType
from slack_channel.abstract_event_handler import AbstractEventHandler
from slack_channel.slack_parser import SlackParser

command_symbol = "reasons"


def _build_response(result, recipient) -> str:
    karma_keys = {str(KarmaType.POZZYPOZ): "++", str(KarmaType.NEGGYNEG): "--"}
    response = "Recent Karma for " + recipient + ":"
    response += os.linesep
    response += "Recently awarded with no reason: " + str(result["reasonless"]) + "."
    karma_with_reasons = result["reasoned"]
    if len(karma_with_reasons) == 0:
        response += os.linesep
        response += "None awarded with a reason given."
        return response

    for k in karma_with_reasons:
        response += os.linesep
        response += karma_keys[k.karma_type] + " "
        response += "from " + k.awarded_by_username + " "
        response += 'for "' + k.reason + '"'

    return response


class GetCurrentKarmaReasonsEventHandler(AbstractEventHandler):
    @property
    def command(self) -> GetCurrentKarmaReasonsCommand:
        return GetCurrentKarmaReasonsCommand()

    @property
    def name(self):
        return "Karma Reasons"

    def _get_command_symbol(self):
        return command_symbol

    def get_usage(self):
        return self.command_trigger + "[recipient <if blank, will default to you>]"

    def can_handle(self, slack_event):
        text = slack_event["text"]
        return text.startswith(self.command_trigger[:-1])  # Trim the space off in case of no recipient

    def _invoke_handler_logic(self, slack_event):
        message = SlackParser.replace_slack_id_tokens_with_usernames(slack_event["text"])
        recipient = message[len(self.command_trigger):].split(' ')[0]

        if recipient == "":
            recipient = slack_event["user"]

        result = self.command.execute(recipient)

        response = _build_response(result, recipient)
        self._send_dm_response(response, slack_event)

File: bottomly.net/Bottomly/Slack/EventHandlers/GetCurrentKarmaReasonsEventHandler.cs
using System.Text;
using Bottomly.Commands;
using Bottomly.Configuration;
using Bottomly.Models;
using Bottomly.Repositories;
using Microsoft.Extensions.Logging;
using Microsoft.Extensions.Options;
using SlackNet.Events;

namespace Bottomly.Slack.EventHandlers;

public class GetCurrentKarmaReasonsEventHandler(
    GetCurrentKarmaReasonsCommand command,
    SlackParser parser,
    ISlackMessageBroker broker,
    IOptions<BottomlyOptions> options,
    ILogger<GetCurrentKarmaReasonsEventHandler> logger)
    : AbstractEventHandler(broker, options, logger)
{
    public override string Name => "Karma Reasons";
    public override ICommand Command => command;
    protected override string CommandSymbol => "reasons";
    public override string GetUsage() => CommandTrigger + "[recipient <if blank, will default to you>]";

    public override bool CanHandle(MessageEvent message) =>
        message.Text?.StartsWith(CommandTrigger.TrimEnd()) == true;

    protected override async Task InvokeHandlerLogicAsync(MessageEvent message)
    {
        var text = await parser.ReplaceSlackIdTokensWithUsernamesAsync(message.Text!);
        var recipient = text[CommandTrigger.Length..].Split(' ')[0];
        if (string.IsNullOrEmpty(recipient))
        {
            recipient = message.User;
        }

        var result = await command.ExecuteAsync(recipient);
        var response = BuildResponse(result, recipient);
        await SendDmResponseAsync(response, message);
    }

    private static string BuildResponse(KarmaReasonsResult result, string recipient)
    {
        var sb = new StringBuilder($"Recent Karma for {recipient}:{Environment.NewLine}");
        sb.Append($"Recently awarded with no reason: {result.Reasonless}.");

        if (!result.Reasoned.Any())
        {
            sb.AppendLine();
            sb.Append("None awarded with a reason given.");
            return sb.ToString();
        }

        foreach (var k in result.Reasoned)
        {
            sb.AppendLine();
            var symbol = k.KarmaType == KarmaType.PozzyPoz ? "++" : "--";
            sb.Append($"{symbol} from {k.AwardedByUsername} for \"{k.Reason}\"");
        }

        return sb.ToString();
    }
}